From 1d4a7ff06ec47acc66645b329e6c2c0264026db5 Mon Sep 17 00:00:00 2001
From: samaga Prasanna Krishna <samaga.krishna@intel.com>
Date: Mon, 24 Sep 2012 16:40:13 +0530
Subject: [PORT FROM MAIN]Last buffer playback issue when seeked to EOS

BZ: 58951

This patch will fix offload playback issue when seeked to end of the
stream and also fixes posting EOS with zero delay.

Change-Id: I01f11709e1d5bd83edc34ee4e3799e79a7adcc11
Orig-Change-Id: I5bc86c248d729869528209a6bc13f89ec53b48ed
origChange-Id: I1123d1fb928c140cabe8fbb4bedd79cd2c142d69
Signed-off-by: samaga Prasanna Krishna <samaga.krishna@intel.com>
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 media/libmediaplayerservice/MediaPlayerService.cpp |    7 ++++++-
 media/libstagefright/AwesomePlayer.cpp             |   16 +++++++++++++---
 2 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/media/libmediaplayerservice/MediaPlayerService.cpp b/media/libmediaplayerservice/MediaPlayerService.cpp
index 14f621b..98b866f 100644
--- a/media/libmediaplayerservice/MediaPlayerService.cpp
+++ b/media/libmediaplayerservice/MediaPlayerService.cpp
@@ -1727,9 +1727,14 @@ void MediaPlayerService::AudioOutput::CallbackWrapper(
     {
         size_t actualSize = (*me->mCallback)(
             me, buffer->raw, buffer->size, me->mCallbackCookie);
-        if (actualSize == 0 && buffer->size > 0 && me->mNextOutput == NULL) {
+        if ( (actualSize == 0) && (buffer->size > 0) &&
+             ((me->mNextOutput == NULL) ||
+             (me->mFlags & AUDIO_OUTPUT_FLAG_COMPRESS_OFFLOAD)) ) {
             // We've reached EOS but the audio track is not stopped yet,
             // keep playing silence.
+            // In offload, buffer size is large and we've not yet finished
+            // writing last buffer to HAL. So keep filling buffer with zero's
+            // to avoid buffer time-out until EOS is posted.
             memset(buffer->raw, 0, buffer->size);
             actualSize = buffer->size;
        }
diff --git a/media/libstagefright/AwesomePlayer.cpp b/media/libstagefright/AwesomePlayer.cpp
index d3c254f..8bcced4 100644
--- a/media/libstagefright/AwesomePlayer.cpp
+++ b/media/libstagefright/AwesomePlayer.cpp
@@ -3393,16 +3393,26 @@ status_t AwesomePlayer::setAACParameters(sp<MetaData> meta, audio_format_t *aFor
  */
 void AwesomePlayer::offloadPauseStartTimer(int64_t time, bool at_pause) {
 #ifdef INTEL_MUSIC_OFFLOAD_FEATURE
-    struct sigevent  pausedEvent;
-    struct itimerspec its;
     ALOGV("offloadPauseStartTimer with time = %lld ", time);
+
     timer_delete(mPausedTimerId);
+
+    if (time == 0) {
+        ALOGV("offloadPauseStartTimer: Posting EOS immediately");
+        mOffloadPostAudioEOS = true;
+        postAudioEOS(0);
+        mOffloadCalAudioEOS= false;
+        return;
+    }
+
+    struct sigevent  pausedEvent;
+    struct itimerspec its;
     memset(&pausedEvent,0, sizeof(sigevent));
     pausedEvent.sigev_notify = SIGEV_THREAD;
+
     if (at_pause) {
         pausedEvent.sigev_notify_function = &timerCallback;
     } else {
-
         pausedEvent.sigev_notify_function = &timerCallbackEOS;
         mOffloadCalAudioEOS= true;
         mOffloadPostAudioEOS = false;
-- 
1.7.4.1

